### DISCUSSION TITLE
URP scene view.

### DISCUSSION THREAD

**Discussion by frettini** [happy-turtle/oit-unity#21] [General]
Hi!
First off thanks for the repo, it's been super useful to learn about OIT :)
I investigated a bit how to do fullscreen blit in URP and here's what I found, mind you I am currently on 2022.3.8. 

In the Execute() function, cmd.Blit() should be replaced by Blitter.BlitCameraTexture(). 
```
public void Render(CommandBuffer command, RenderTargetIdentifier src, RenderTargetIdentifier dest)
  {
      command.ClearRandomWriteTargets();
     if(linkedListMaterial== null 
     || renderingData.cameraData.isPreviewCamera)
     {
         return;
     }

      // blend linked list
      linkedListMaterial.SetBuffer(fragmentLinkBufferId, fragmentLinkBuffer);
      linkedListMaterial.SetBuffer(startOffsetBufferId, startOffsetBuffer);
       Blitter.BlitCameraTexture(cmd, 
       src, 
       dest, 
       linkedListMaterial, 0);
  }
```
I'm checking for null material because reopening the scene made it null for some reason. Checking for isPreviewCamera is used to avoid some asserts in the Blitter when selecting objects in the scene.

I also slightly changed OitFullscreenRender.shader to use the _CameraOpaqueTexture and by including Blit.hlsl. This also requires to remove the attributes, varyings, and vertex shader as they are already defined in Blit.hlsl.

```
 // The Blit.hlsl file provides the vertex shader (Vert),
// input structure (Attributes) and output strucutre (Varyings)
#include "Packages/com.unity.render-pipelines.core/Runtime/Utilities/Blit.hlsl"

 TEXTURE2D_X(_CameraOpaqueTexture);
 SAMPLER(sampler_CameraOpaqueTexture);
``` 
which could later then be sampled as such 
```
 half3 col = SAMPLE_TEXTURE2D_X(_CameraOpaqueTexture, sampler_CameraOpaqueTexture, input.texcoord).rgb;
```
and that seemed to do the trick for me. Most of it comes from https://docs.unity3d.com/Packages/com.unity.render-pipelines.universal@14.0/manual/renderer-features/how-to-fullscreen-blit.html . Hopefully that applies to your version as well!

**Comment by happy-turtle** [comment]
Hi, thanks a lot for your input! That is really helpful. And I am glad it was of some use. 😊 If you are able to and you have the code ready anyways, it would be nice if you could start a pull request with these changes. And I will pick up on it. 

Either way I will try to find the time for this. The Unity API is unfortunately very different between the pipelines which makes it extra hard to get into it as I am currently not using it on a daily basis.

**Comment by happy-turtle** [comment]
https://github.com/happy-turtle/oit-unity/pull/22

**Comment by frettini** [comment]
Hi I'll try to have a look at it in the coming days :) Also please bear with me as this would be my first pull request.

**Reply by happy-turtle** [reply]
No worries, I don't expect a perfect solution. Just glad for any help

**Comment by happy-turtle** [comment]
I've been trying to set it up here https://github.com/happy-turtle/oit-unity/commit/0b3012b4c2561036136d504acc107e0afe456a7c
But I am having a bit of trouble. Using the Blitter works fine, and actually HDRP works with it. But updating the Fullscreen shader for URP is giving me shader conflicts when I try to import Blit.hlsl. Will see if I can figure this out. Maybe you have a hint on this?

**Reply by frettini** [reply]
Hey, nice one for hdrp! What kind of conflicts do you get? I was using 
 #include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/Core.hlsl"
 #include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/UnityInput.hlsl"
 instead of #include "UnityCG.cginc", but that means that it will only work in URP. 
 
I've tried to run a fork on 2021.3.9 as this is the last tested version in the readme but I couldn't even get to open a project in that version. I'm now using 2021.3.33 (LTS) but the render pass doesn't work out of the box. The URP version on this one is 12.1.13 whereas the one I got it to work on was 14.x.x, and the API varries quite a lot in between unfortunately.

**Reply by happy-turtle** [reply]
I am either getting TEXTURE_2_X not defined or something about parameters being defined multiple times. I will try out the include file you mentioned. I updated to Unity 2022 though. I would say this makes sense as the APIs for the new render pipelines seem to be a better and it is the latest LTS version. It is a bit annoying though that the APIs change so quickly. Will have to update the minimum package version then.

**Comment by happy-turtle** [comment]
By now I got it working on the feature branch, that's really nice. I will need to split the fullscreen shader into multiple shaders to still work with all of the pipelines though. Writing shaders for all pipelines is a pain at the moment. They are working on it, but it will take more time https://forum.unity.com/threads/block-shaders-surface-shaders-for-srps-and-more-public-demo-now-available.1350497/.

Feeling good about getting this done at some point though. Thanks again for your input 😊

**Reply by frettini** [reply]
Nice one! Thanks for this :))
Hopefully this block shaders feature comes soon then :)

**Reply by frettini** [reply]
Hello, hope you are well! 
I just came across this interesting comment in https://github.com/ColinLeung-NiloCat/UnityURPToonLitShaderExample/blob/master/SimpleURPToonLitOutlineExample.shader : 
```
Tags 
        {
            // SRP introduced a new "RenderPipeline" tag in Subshader. This allows you to create shaders
            // that can match multiple render pipelines. If a RenderPipeline tag is not set it will match
            // any render pipeline. In case you want your SubShader to only run in URP, set the tag to
            // "UniversalPipeline"

            // here "UniversalPipeline" tag is required, because we only want this shader to run in URP.
            // If Universal render pipeline is not set in the graphics settings, this SubShader will fail.

            // One can add a SubShader below or fallback to Standard built-in to make this
            // material works with both Universal Render Pipeline and Builtin-RP

            // the tag value is "UniversalPipeline", not "UniversalRenderPipeline", be careful!
            "RenderPipeline" = "UniversalPipeline"

            // explicit SubShader tag to avoid confusion
            "RenderType" = "Opaque"
            "IgnoreProjector" = "True"
            "UniversalMaterialType" = "ComplexLit"
            "Queue"="Geometry"
        }
```
not ideal, but it could maybe help have different shaders for different pipelines..

**Reply by happy-turtle** [reply]
Thanks for the reference, it's a lot more explanatory than what I have found so far. I briefly tried to use this when adding the URP scene view, but couldn't get it to work. It also seems the tags are documented differently between Unity versions. 😅 Sometimes it's "UniversalPipeline" and sometimes "UniversalRenderPipeline". But I will have a look at this again and see if I can simplify the shaders. This will get easier as the Scriptable Render Pipeline is maturing.

Also as we are at it: Do you maybe know of a good way to reference Shader assets on e.g. a URP RenderPass? Currently I need to put all of the fullscreen shaders into a Resources folder so that they are included in build. I basically need a default asset reference that I can assign in the package, so that a user would not need to set the reference by themself in the Editor. A nuisance that is hard to explain 😄 Basically instead of passing a string as the shader name [here](https://github.com/happy-turtle/oit-unity/blob/863657d0ad4971bb71a28f5b50403896c7e91c36/URP/Runtime/OitPass.cs#L15C78-L15C78) I would like to pass in the actual shader asset reference. But a RenderPass is no MonoBehaviour so I am not able to assign the reference in the Editor.

**Reply by frettini** [reply]
I get what you mean. A lot of tutorials rely on the user adding the material or shader reference to the pass manually, which isn't great. I've done some investigation but couldn't find a way to get a default reference. The only thing I found is the use of the `Reload` attribute in ScreenSpaceAmbientOcclusion.cs : 
```
[SerializeField]
[HideInInspector]
[Reload("Shaders/Utils/ScreenSpaceAmbientOcclusion.shader")]
private Shader m_Shader;
```
For this to work, it also needs the following lines in the `Create()` method :
```
        public override void Create()
        {
#if UNITY_EDITOR
            ResourceReloader.TryReloadAllNullIn(this, UniversalRenderPipelineAsset.packagePath);
#endif
```
The material is then created using `mat = CoreUtils.CreateEngineMaterial(shader);`. On one hand, it removes the need for a resource folder, but you will still need to write the shader name and path for this to work, so it's not as good as a reference.

**Reply by happy-turtle** [reply]
That's a cool find, thanks for looking into it! I will try that out when I have time. The only downside is that this is using the SRP API. But maybe there is an alternative for the Bultin pipeline.

**Comment by happy-turtle** [comment]
Implemented with https://github.com/happy-turtle/oit-unity/pull/22
